### PR TITLE
deps: Update controller-runtime to v0.16.3, and other packages to latest versions

### PR DESCRIPTION
**Pull request by akrejcir**
**What this PR does / why we need it**:
Updated:
- Golang version to `1.20`.
- `github.com/openshift/*` modules to branch `release-4.15`.
- `github.com/prometheus/*` modules to latest versions.
- `github.com/tektoncd/pipeline` to `v0.54.0`.
- `k8s.io/*` modules to `v0.28.3`.
- `kubevirt.io/api` to `v1.1.0`.
- `kubevirt.io/containerized-data-importer-api` to `v1.58.0`.
- `sigs.k8s.io/controller-runtime` to `v0.16.3`.
- `sigs.k8s.io/kustomize/*` modules to latest versions.

**Which issue(s) this PR fixes**: 
Fixes # https://github.com/kubevirt/ssp-operator/issues/597

**Release note**:
```release-note
None
```

### CODE REVIEW

**Comment by akrejcir** [comment]
This is ready for review.

/cc @0xFelix @ksimon1

File: webhooks/ssp_webhook_test.go [hunk 1]
```diff
@@ -144,7 +144,8 @@ var _ = Describe("SSP Validation", func() {
 				},
 			}
 
-			Expect(validator.ValidateCreate(ctx, toUnstructured(ssp))).To(Succeed())
+			_, err := validator.ValidateCreate(ctx, toUnstructured(ssp))
+			Expect(err).To(Succeed())
```
Comment (0xFelix) [inline webhooks/ssp_webhook_test.go:148]: nit: ToNot(HaveOccured())

File: webhooks/ssp_webhook_test.go [hunk 2]
```diff
@@ -239,17 +248,20 @@ var _ = Describe("SSP Validation", func() {
 
 		It("should reject URL without https:// or ssh://", func() {
 			sspObj.Spec.CommonInstancetypes.URL = pointer.String("file://foo/bar")
-			Expect(validator.ValidateCreate(ctx, toUnstructured(sspObj))).ShouldNot(Succeed())
+			_, err := validator.ValidateCreate(ctx, toUnstructured(sspObj))
+			Expect(err).ShouldNot(Succeed())
```
Comment (0xFelix) [inline webhooks/ssp_webhook_test.go:252]: I'd go for one style throughout the whole file, either `HaveOccured()` or `Succeed()`.

File: internal/common/resource.go [hunk 3]
```diff
@@ -529,7 +529,7 @@ func defaultStatusFunc(obj client.Object) ResourceStatus {
 			msg := fmt.Sprintf("No pods for deployment %s/%s are running. Expected: %d",
 				objTyped.Namespace,
 				objTyped.Name,
-				objTyped.Status.Replicas)
+				numberOfReplicas)
```
Comment (0xFelix) [inline internal/common/resource.go:532]: nit: Looks like this change belongs into another commit?zz

File: internal/common/resource.go [hunk 4]
```diff
@@ -499,8 +499,8 @@ func defaultUpdateFunc(newObj, foundObj client.Object) {
 	case *promv1.ServiceMonitor:
 		foundObj.(*promv1.ServiceMonitor).Spec = newTyped.Spec
 
-	case *tekton.Task:
-		foundObj.(*tekton.Task).Spec = newTyped.Spec
+	case *tekton.Task: //nolint:staticcheck
```
Comment (0xFelix) [inline internal/common/resource.go:502]: Can you explain why `//nolint:staticcheck` is necessary?

File: internal/common/environment.go [hunk 5]
```diff
@@ -63,8 +65,9 @@ func RunningOnOpenshift(ctx context.Context, cl client.Reader) (bool, error) {
 			Name: "version",
 		},
 	}
-	if err := cl.Get(ctx, client.ObjectKeyFromObject(clusterVersion), clusterVersion); err != nil {
-		if meta.IsNoMatchError(err) || apierrors.IsNotFound(err) {
+	err := cl.Get(ctx, client.ObjectKeyFromObject(clusterVersion), clusterVersion)
```
Comment (0xFelix) [inline internal/common/environment.go:68]: nit: Keep inlining of `err`?

File: .golangci.yaml [hunk 6]
```diff
@@ -1,3 +1,6 @@
+run:
+  modules-download-mode: readonly
```
Comment (0xFelix) [inline .golangci.yaml:2]: Why is it necessary?

File: api/go.mod [hunk 7]
```diff
@@ -1,19 +1,19 @@
...
 
-go 1.19
+go 1.20
 
 require (
-	github.com/openshift/api v0.0.0-20230228142948-d170fcdc0fa6 // release-4.13
-	k8s.io/apimachinery v0.26.10
-	kubevirt.io/containerized-data-importer-api v1.55.2
+	github.com/openshift/api v0.0.0-20231130063041-be20875a1373 // release-4.15
+	k8s.io/apimachinery v0.28.3
```
Comment (0xFelix) [inline api/go.mod:7]: There is `v0.28.4` already, can we use it?

File: internal/operands/tekton-pipelines/reconcile.go [hunk 8]
```diff
@@ -49,7 +49,9 @@ func WatchClusterTypes() []operands.WatchType {
 }
 
 type tektonPipelines struct {
-	pipelines       []pipeline.Pipeline
+	// TODO: Update to v1.Pipeline
```
Comment (0xFelix) [inline internal/operands/tekton-pipelines/reconcile.go:52]: Can you open an issue for it?

**Review by 0xFelix — COMMENTED** [review]
Great work! Added some comments.

File: .golangci.yaml (hunk 6 above)
Comment (akrejcir) [reply .golangci.yaml:2]: It may not be necessary anymore. I was getting linter errors in CI, but not locally, and after googling possible cause I found that it may be trying to update vendored modules.

The issue may also have been caused by using golang 1.19 in CI, when the code needs 1.20.
I will try to remove it.

File: internal/common/resource.go (hunk 3 above)
Comment (akrejcir) [reply internal/common/resource.go:532]: I will move it to another PR.

File: webhooks/ssp_webhook_test.go (hunk 1 above)
Comment (akrejcir) [reply webhooks/ssp_webhook_test.go]: done

File: webhooks/ssp_webhook_test.go (hunk 2 above)
Comment (akrejcir) [reply webhooks/ssp_webhook_test.go]: done

File: internal/common/environment.go (hunk 5 above)
Comment (akrejcir) [reply internal/common/environment.go]: done

File: api/go.mod (hunk 7 above)
Comment (akrejcir) [reply api/go.mod]: done

File: internal/common/resource.go (hunk 3 above)
Comment (0xFelix) [reply internal/common/resource.go]: ACK

File: internal/common/resource.go (hunk 4 above)
Comment (0xFelix) [reply internal/common/resource.go:502]: ACK

File: internal/crd-watch/crd-watch_test.go [hunk 9]
```diff
@@ -9,7 +9,7 @@ import (
 	. "github.com/onsi/ginkgo/v2"
 	. "github.com/onsi/gomega"
 
-	"k8s.io/apiextensions-apiserver/pkg/apis/apiextensions"
+	apiextensions "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
```
Comment (0xFelix) [inline internal/crd-watch/crd-watch_test.go:12]: Same as above

File: internal/operands/common-instancetypes/reconcile_test.go [hunk 10]
```diff
@@ -9,14 +9,14 @@ import (
 	. "github.com/onsi/ginkgo/v2"
 	. "github.com/onsi/gomega"
 
-	"k8s.io/apiextensions-apiserver/pkg/apis/apiextensions"
+	apiextensions "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
```
Comment (0xFelix) [inline internal/operands/common-instancetypes/reconcile_test.go:12]: Same as above

File: internal/operands/tekton-tasks/reconcile_test.go [hunk 11]
```diff
@@ -9,7 +9,7 @@ import (
 	pipeline "github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1"
 	v1 "k8s.io/api/core/v1"
 	rbac "k8s.io/api/rbac/v1"
-	"k8s.io/apiextensions-apiserver/pkg/apis/apiextensions"
+	apiextensions "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
```
Comment (0xFelix) [inline internal/operands/tekton-tasks/reconcile_test.go:12]: Same as above

Comment (akrejcir) [reply internal/operands/tekton-tasks/reconcile_test.go]: done

File: internal/operands/common-instancetypes/reconcile_test.go (hunk 10 above)
Comment (akrejcir) [reply internal/operands/common-instancetypes/reconcile_test.go]: done

File: internal/crd-watch/crd-watch_test.go (hunk 9 above)
Comment (akrejcir) [reply internal/crd-watch/crd-watch_test.go]: done

File: internal/crd-watch/crd-watch.go [hunk 12]
```diff
@@ -5,13 +5,12 @@ import (
 	"fmt"
 	"sync"
 
-	"k8s.io/apiextensions-apiserver/pkg/apis/apiextensions"
+	apiextensions "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
```
[the remark this answers is not in the document]
Comment (akrejcir) [reply internal/crd-watch/crd-watch.go]: doen

**Review by jcanocan — COMMENTED** [review]
Nice job @akrejcir !

File: tests/tests_suite_test.go [hunk 13]
```diff
@@ -9,13 +9,15 @@ import (
...
 	. "github.com/onsi/gomega"
+
+	ginkgo_reporters "github.com/onsi/ginkgo/v2/reporters"
 	osconfv1 "github.com/openshift/api/config/v1"
 	routev1 "github.com/openshift/api/route/v1"
 	secv1 "github.com/openshift/api/security/v1"
 	templatev1 "github.com/openshift/api/template/v1"
 	promv1 "github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring/v1"
+	pipeline "github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1"
 	apps "k8s.io/api/apps/v1"
```
[the remark this answers is not in the document]
Comment (akrejcir) [reply tests/tests_suite_test.go:21]: No. Imports are usually separated into 3 blocks:
- golang standard library
- imports from modules in `go.mod`
- local imports from this project

For tests there is a 4th block:
- Dot imports of testing modules like ginkgo and gomega.

[the remark this answers is not in the document]
Comment (jcanocan) [reply tests/tests_suite_test.go:21]: Great! Thanks for letting me know :)

**Review by 0xFelix — COMMENTED** [review]
/approve

**Comment by akrejcir** [comment]
The functional tests pass locally.

**Comment by jcanocan** [comment]
Recreated this PR since it requires this change https://github.com/kubevirt/ssp-operator/pull/792 to pass CI. Without it, it  CI won't work. Hope you don't mind @akrejcir.

**Comment by 0xFelix** [comment]
Superseded by https://github.com/kubevirt/ssp-operator/pull/795
